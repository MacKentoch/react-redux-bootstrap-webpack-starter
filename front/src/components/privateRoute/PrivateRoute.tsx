import React from 'react';
import { Route, Redirect } from 'react-router-dom';
import { RouteComponentProps } from 'react-router';
import { ReduxConnectedProps, OwnProps } from './index';

// #region types
type Props = RouteComponentProps & OwnProps & ReduxConnectedProps;
// #endregion

function PrivateRoute(props: Props) {
  const { children: InnerComponent, ...rest } = props;
  const { location, isAuthenticated } = props;

  return (
    <Route
      {...rest}
      render={() =>
        isAuthenticated ? (
          InnerComponent
        ) : (
          <Redirect to={{ pathname: '/login', state: { from: location } }} />
        )
      }
    />
  );
}

PrivateRoute.displayName = 'PrivateRoute';

export default PrivateRoute;
