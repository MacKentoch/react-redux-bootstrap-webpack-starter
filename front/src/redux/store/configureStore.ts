import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { createBrowserHistory as createHistory } from 'history';
import { persistReducer, persistStore } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import thunkMiddleware from 'redux-thunk';
import { createLogger } from 'redux-logger';
import reducer from '../modules/reducers';

// #region constants
const isProd = process.env.NODE_ENV === 'production';
export const history = createHistory();
// #endregion

// #region createStore : enhancer

// #region logger middleware (dev only)
const loggerMiddleware = createLogger({
  level: 'info',
  collapsed: true,
});
// #endregion

const enhancer = !isProd
  ? composeWithDevTools(
      applyMiddleware(
        thunkMiddleware,
        loggerMiddleware, // logger at the end
      ),
    )
  : composeWithDevTools(applyMiddleware(thunkMiddleware));
// #endregion

// #region persisted reducer
const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['router'],
  // whitelist: ['userAuth'],
};

const persistedReducer = persistReducer(persistConfig, reducer);
// #endregion

export default function configureStore(initialState = {}) {
  const store = createStore(persistedReducer, initialState, enhancer);
  const persistor = persistStore(store);

  if ((module as any).hot) {
    (module as any).hot?.accept('../modules/reducers', () => {
      // @ts-ignore
      const nextAppReducer = require('../modules/reducers').default;
      store.replaceReducer(persistReducer(persistConfig, nextAppReducer));
    });
  }

  return { store, persistor };
}
