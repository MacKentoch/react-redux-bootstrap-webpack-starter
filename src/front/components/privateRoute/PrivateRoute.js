// @flow

// #region imports
import React, { Component } from 'react';
import { Route, Redirect, withRouter } from 'react-router-dom';
import {
  type Match,
  type Location,
  type RouterHistory,
} from 'react-router-dom';
import auth from '../../services/auth';
// #endregion

// #region flow types
type Props = {
  // react-router 4:
  match: Match,
  location: Location,
  history: RouterHistory,

  // parent
  component: any,
  path: string,

  ...any,
};
type State = any;
// #endregion

class PrivateRoute extends Component<Props, State> {
  // #region lifecycle
  render() {
    const { component: InnerComponent, ...rest } = this.props;
    const { location } = this.props;

    const isUserAuthenticated = this.isAuthenticated();
    const isTokenExpired = false; // this.isExpired();

    return (
      <Route
        {...rest}
        render={props =>
          !isTokenExpired && isUserAuthenticated ? (
            <InnerComponent {...props} />
          ) : (
            <Redirect to={{ pathname: '/login', state: { from: location } }} />
          )
        }
      />
    );
  }
  // #endregion

  isAuthenticated() {
    const checkUserHasId = user => user && user.id;
    const user = auth.getUserInfo() ? auth.getUserInfo() : null;
    const isAuthenticated = auth.getToken() && checkUserHasId(user);
    return isAuthenticated;
  }

  isExpired() {
    return auth.isExpiredToken(auth.getToken());
  }
}

export default withRouter(PrivateRoute);
