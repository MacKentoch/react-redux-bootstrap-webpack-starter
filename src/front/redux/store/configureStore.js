// @flow

/* eslint no-process-env:0 */
if (process.env.NODE_ENV === 'production') {
  module.exports = require('./configureStore.prod');
} else {
  module.exports = require('./configureStore.dev');
}
