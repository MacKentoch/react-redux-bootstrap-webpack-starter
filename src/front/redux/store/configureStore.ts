/* eslint no-process-env:0 */
let configureStore = null;

if (process.env.NODE_ENV === 'production') {
  configureStore = require('./configureStore.prod').default;
  // const {history} = require('./configureStore.prod');
  // export history;
} else {
  configureStore = require('./configureStore.dev').default;
  // const {history} = require('./configureStore.dev');
  // export history;
}
export default configureStore;
