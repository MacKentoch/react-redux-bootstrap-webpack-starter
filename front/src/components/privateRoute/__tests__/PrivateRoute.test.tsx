import React from 'react';
import { render } from '@testing-library/react';
import { Router, Switch, useHistory } from 'react-router';
import { Route } from 'react-router';
import { createHashHistory } from 'history';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import configureStore from 'redux-mock-store';
import PrivateRoute from '../index';

// #region constants
const history = createHashHistory();
const middlewares: Array<any> = [];
const mockStore = configureStore(middlewares);
// #endregion

describe('PrivateRoute component', () => {
  let rootElement: any = null;

  const HomePage = () => {
    // const history = useHistory();
    history.push('/protected');
    return <p>Home page</p>;
  };

  beforeEach(() => {
    rootElement = document.createElement('div');
    document.body.appendChild(rootElement);
  });

  afterEach(() => {
    rootElement && document.body.removeChild(rootElement);
    rootElement = null;
  });

  it('renders as expected', async () => {
    const initialState = {
      userAuth: {
        isFetching: false,
        isLogging: false,
        id: '',
        login: '',
        firstname: '',
        lastname: '',
        token: '',
        isAuthenticated: false,
      },
    };
    const store = mockStore(initialState);
    const props = {};

    const LoginPage = () => <p>Login page</p>;
    const ChildPage = () => <p>private page</p>;

    const { container } = render(
      <Provider store={store}>
        <ThemeProvider theme={{}}>
          <Router history={history}>
            <Switch>
              <Route exact path="/">
                <HomePage />
              </Route>
              <PrivateRoute {...props} path="/protected">
                <ChildPage />
              </PrivateRoute>
              <Route exact path="/login">
                <LoginPage />
              </Route>
            </Switch>
          </Router>
        </ThemeProvider>
      </Provider>,
      rootElement,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('redirects to login when not authenticated', async () => {
    const initialState = {
      userAuth: {
        isFetching: false,
        isLogging: false,
        id: '',
        login: '',
        firstname: '',
        lastname: '',
        token: '',
        isAuthenticated: false,
      },
    };
    const store = mockStore(initialState);
    const props = {};
    const PrivatePage = () => <p data-testid="private">private page</p>;
    const LoginPage = () => <p data-testid="login">login page</p>;

    const { findByTestId } = render(
      <Provider store={store}>
        <ThemeProvider theme={{}}>
          <Router history={history}>
            <Switch>
              <Route exact path="/">
                <HomePage />
              </Route>
              <PrivateRoute {...props} path="/protected">
                <PrivatePage />
              </PrivateRoute>
              <Route exact path="/login">
                <LoginPage />
              </Route>
            </Switch>
          </Router>
        </ThemeProvider>
      </Provider>,
      rootElement,
    );

    const loginPageContainer = await findByTestId('private');
    expect(loginPageContainer.textContent).toBe('private page');
  });
});
