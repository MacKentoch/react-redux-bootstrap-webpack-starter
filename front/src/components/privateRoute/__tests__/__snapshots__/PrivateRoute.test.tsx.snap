// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute component renders as expected 1`] = `
<p>
  private page
</p>
`;
